src/main/java/org/myrobotlab/codec/Api.java

```diff
@@ -142,21 +158,214 @@ static public Message uriToMsg(String uri) {
     return msg;
   }
 
-  static public String getApiKey(String requestUri) {
+  /**
+   * needed to get the api key to select the appropriate api processor
+   * @param r
+   * @return
+   */
+  static public String getApiKey(AtmosphereResource r) {
+    String requestUri = r.getRequest().getRequestURI();
+    return getApiKey(requestUri);
+  }
 
-    int pos = requestUri.indexOf(Api.PARAMETER_API);
+  static public String getApiKey(String uri) {
+    int pos = uri.indexOf(Api.PARAMETER_API);

```

Comment:
this seems a little fragile... ok for now..  we might want to consider a URL class to handle this parsing for us.

Reply:
URL class (and URI) handle parsing of the full url, but nothing I know of in the jvm will handle "parts" of the path.  Mrl apis are very structured in that the api is determined by the path-part after the constant "/api/"
e.g. "/api/"{api-type}